optuna/samplers/_tpe/sampler.py

```diff
@@ -835,9 +835,17 @@ def _feasible(trial: FrozenTrial) -> bool:
 
     loo_mat = ~np.eye(pareto_sols.shape[0], dtype=bool)  # Leave-one-out bool matrix.
     contribs = np.zeros(n_below_feasible, dtype=float)
-    contribs[on_front] = hv - np.array(
-        [compute_hypervolume(pareto_sols[loo], ref_point, assume_pareto=True) for loo in loo_mat]
-    )
+    if len(study.directions) <= 3:
+        contribs[on_front] = [
+            hv - compute_hypervolume(pareto_sols[loo], ref_point, assume_pareto=True)
+            for loo in loo_mat
+        ]
+    else:
+        contribs[on_front] = np.prod(ref_point - pareto_sols, axis=-1)
+        limited_sols = np.maximum(pareto_sols, pareto_sols[:, np.newaxis])
+        contribs[on_front] -= [
+            compute_hypervolume(limited_sols[i, loo], ref_point) for i, loo in enumerate(loo_mat)
+        ]
```

Comment:
Just in case, I confirmed that this part does not fail even if `len(pareto_sols) == 1`.
The hypervolume contribution in this case must be identical to `compute_hypervolume(pareto_sols, ref_point, assume_pareto=True)`, which we successfully get here.

> [!NOTE]
> Since `n_below_feasible <= 1` immediately returns `weights_below` above, `pareto_sols` will not be empty.